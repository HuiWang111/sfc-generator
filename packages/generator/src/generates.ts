import type { GeneratorOptions } from '@babel/generator'
import type { SFCBlock } from '@vue/compiler-sfc'
import type {
  BabelNode,
  GenerateOptions,
  ScriptAttrs,
  SFCGenerateOptions,
  TemplateNode,
  TextNode,
} from './types'
import generate from '@babel/generator'
import { TemplateNodeType } from './constants'
import { Template } from './template'

/**
 * 用于生成 template 代码的函数
 * @param node template 代码的 ast
 * @param options 生成配置
 * @returns template 代码
 */
export function generateTemplate(node: TemplateNode, options: GenerateOptions = {}) {
  const template = new Template(options)
  const content = template.generate(node)
  return ['<template>', content, '</template>'].join('\n')
}

/**
 * 用于生成 script 代码的函数
 * @param node babel 的 ast node
 * @param attrs script 标签的属性
 * @param opts 生成配置
 * @returns script 代码
 */
export function generateScript(
  node: BabelNode,
  attrs: ScriptAttrs = {},
  opts: GeneratorOptions = {},
) {
  const { code } = generate(node, opts)
  const template = new Template({
    autoIndent: false,
  })
  const scriptContentNode: TextNode = {
    type: TemplateNodeType.Text,
    text: code,
  }
  const attrsMap = Object.entries(attrs).reduce<Record<string, string>>((acc, [key, value]) => {
    if (value != null) {
      acc[key] = value
    }
    return acc
  }, {})

  return template.generate(
    {
      type: TemplateNodeType.Element,
      tag: 'script',
      attrsMap,
      children: [scriptContentNode],
    },
    0,
  )
}

/**
 * 用于生成 style 代码的函数
 * @param styles 生成 style 的 SFCBlock
 * @returns style 的代码
 */
export function generateStyle(styles: SFCBlock[]) {
  if (!styles.length)
    return ''

  const template = new Template({
    autoIndent: false,
  })

  return styles.map((style) => {
    const attrsMap: Record<string, any> = {}
    if (style.lang) {
      attrsMap.lang = style.lang
    }
    if (style.scoped) {
      attrsMap.scoped = style.scoped
    }
    const styleContentNode: TextNode = {
      type: TemplateNodeType.Text,
      text: style.content.trim(),
    }

    return template.generate(
      {
        type: TemplateNodeType.Element,
        tag: 'style',
        attrsMap,
        children: [styleContentNode],
      },
      0,
    )
  }).join('\n\n')
}

/**
 * 用于生成单文件组件代码的函数
 * @param opts 生成配置
 * @returns 单文件组件的代码
 */
export function generateComponent(opts: SFCGenerateOptions) {
  const list: string[] = []

  list.push(generateTemplate(opts.template.node, opts.template.options))

  if (opts.script) {
    list.push(generateScript(opts.script.node, opts.script.attrs, opts.script.options))
  }
  if (opts.styles && opts.styles.length) {
    list.push(generateStyle(opts.styles))
  }

  return list.join('\n\n')
}
