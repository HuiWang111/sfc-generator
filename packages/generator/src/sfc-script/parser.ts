import type { ParserOptions } from '@babel/parser'
import type { ScriptParseOptions } from '../types'
import { parse as babelParse } from '@babel/parser'
import traverse from '@babel/traverse'
import { CompositionApi } from './composition-api'
import { OptionsApi } from './options-api'

export function parse<T extends boolean = false>(
  code: string,
  options: ScriptParseOptions<T> = {},
): {
    api: T extends true ? CompositionApi : OptionsApi
    ast: ReturnType<typeof babelParse>
  } {
  const { lang, jsx = false, setup = false } = options

  const plugins: ParserOptions['plugins'] = []
  if (lang === 'ts') {
    plugins.push('typescript')
  }
  if (jsx) {
    plugins.push('jsx')
  }

  const ast = babelParse(code, {
    sourceType: 'module',
    plugins,
  })

  let api: OptionsApi | CompositionApi | null = null

  if (setup) {
    api = new CompositionApi(ast.program.body)
  }
  else {
    traverse(ast, {
      ExportDefaultDeclaration(path) {
        if (setup)
          return

        const declarationPath = path.get('declaration')
        if (declarationPath.isObjectExpression()) {
          api = new OptionsApi(declarationPath)
        }
      },
    })
  }

  if (!api) {
    throw new Error(`script setup attr is ${setup}, parse code error: ${code}`)
  }

  return {
    api: api as any,
    ast,
  }
}
