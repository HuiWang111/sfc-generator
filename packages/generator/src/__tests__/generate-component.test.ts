import type { TemplateNode } from '../types'
import { parse } from '@babel/parser'
import { compileTemplate, parseComponent } from '@vue/compiler-sfc'
import { generateComponent } from '../generates'

describe('test generateComponent', () => {
  const template = `<template>
  <div>
    <n-btn
      style="margin-bottom: 10px;"
      @click="toggleLocale"
    >
      {{ localeCode.includes('zh') ? '切换语言' : 'toggle language' }}
    </n-btn>
    <n-config-provider
      :locale="locale"
    >
      <n-empty />
    </n-config-provider>
  </div>
</template>

<script setup>
import { computed, onMounted, ref } from 'vue'
import { en, zhCN } from '../../locale'

const localeCode = ref('zh-cn')
const locale = computed(() => localeCode.value.includes('zh') ? zhCN : en)

const toggleLocale = () => {
  localeCode.value = localeCode.value.includes('zh')
    ? 'en'
    : 'zh'
}

onMounted(() => {
  localeCode.value = window.navigator.language
})
</script>`

  it('generateComponent should work', () => {
    const sfcDescriptor = parseComponent(template)
    const script = sfcDescriptor.script || sfcDescriptor.scriptSetup
    if (!script)
      return

    const ast = parse(script.content, {
      sourceType: 'module',
    })
    const code = generateComponent({
      template: {
        node: compileTemplate({
          source: sfcDescriptor.template!.content,
          filename: '',
        }).ast as TemplateNode,
      },
      script: {
        node: ast,
        attrs: {
          setup: script.setup as boolean,
        },
      },
      styles: sfcDescriptor.styles,
    })

    expect(code).toMatchSnapshot()
  })
})
