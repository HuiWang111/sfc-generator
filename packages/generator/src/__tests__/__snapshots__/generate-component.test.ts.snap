// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test generateComponent generateComponent should work 1`] = `
"<template>
  <div>
    <n-btn style="margin-bottom: 10px;" @click="toggleLocale">
      {{ localeCode.includes('zh') ? '切换语言' : 'toggle language' }}
    </n-btn>
    <n-config-provider :locale="locale">
      <n-empty />
    </n-config-provider>
  </div>
</template>

<script setup>
import { computed, onMounted, ref } from 'vue';
import { en, zhCN } from '../../locale';

const localeCode = ref('zh-cn');
const locale = computed(() => localeCode.value.includes('zh') ? zhCN : en);

const toggleLocale = () => {
  localeCode.value = localeCode.value.includes('zh') ?
  'en' :
  'zh';
};

onMounted(() => {
  localeCode.value = window.navigator.language;
});
</script>"
`;
